server_environment_data_encryption/readme/CONFIGURE.rst

```diff
@@ -0,0 +1,23 @@
+In order to use this module properly, each environment should have their own encryption key
+and the production environment should have the keys of all environments.
+
+Example :
+Development environment ::
+
+  [options]
+  running_env=dev
+  encryption_key_dev=XXX
+
+Pre-production environment ::
+
+  [options]
+  running_env=preprod
+  encryption_key_preprod=YYY
+
+Production environment ::
+
+  [options]
+  running_env=prod
+  encryption_key_dev=XXX
+  encryption_key_preprod=YYY
+  encryption_key_prod=ZZZ
```

Comment:
I think there's an important step missing here: how does it works?
How am I supposed to provide secret configuration for my records?

By reading the code I understand that when you install it is going to convert env values to encrypted.data records (one per each record). Correct?

But then, how am I supposed to provide my secrets via env keys?
I mean, can I achieve something like I would do w/ `SERVER_ENV_CONFIG_SECRET`?

My guess: you cannot. You rely on manual configuration of the values via UI and simply store them encrypted.
Which means: at first setup, you'll have to configure manually each param on each env.
Unless you generate encrypted.data records w/ some black magic upgrade step/hook.

It would be nice if someone can document this in the README :pray:

Reply:
If I understand correctly, you think we should add a post install hook to encrypt all env dependent clear data inf the database ?
If yes, you are right, there are no such thing yet, and it could be a nice addition. Still it may be tricky because nothing prevent you to install the module without having configured the entryption key yet...

So maybe it would be best to propose a shell script or something someone could run manually if needed it its situation.
But anyway, if you install the module on existing database with password in clear, yes they still will be in clear and it does not change anything. Only editing it again will encrypt it.

For  the how does it works, it is quite simple : 
Each time someone write an env dependent fields, instead of storing it a field in clear, Odoo will read all env dependent values already stored/encrytpted, update it with the modified value and re-encrypt it.

I am not sure about you question about secret configuration, it has to be in the odoo config file like stated in the readme.

Reply:
> If I understand correctly, you think we should add a post install hook to encrypt all env dependent clear data inf the database ?

Nope, I think you should mention this and all the other things you mentioned in this answer that today are not very clear in the README. Then anything else (scripts, hooks, whatever) should stay in a ROADMAP section.

> If yes, you are right, there are no such thing yet, and it could be a nice addition. Still it may be tricky because nothing prevent you to install the module without having configured the entryption key yet...

> So maybe it would be best to propose a shell script or something someone could run manually if needed it its situation.
> But anyway, if you install the module on existing database with password in clear, yes they still will be in clear and it does not change anything. Only editing it again will encrypt it.

Ok, this should definitely be written in the README :wink: 

Regarding the script/hook: I'm not talking about an install hook just for this module (which would be useful and better anyway). I'm talking about new configuration to be provided automatically by custom modules. 

If I add a new param that should be encrypted I have no other mean - as of today - to create it straight from my module (or any other tool). Thus, I have to rely only on manual edit via UI and I cannot ship my devs in a secure way ready to be deployed on all envs. Correct?

> Each time someone write an env dependent fields, instead of storing it a field in clear, Odoo will read all env dependent values already stored/encrytpted, update it with the modified value and re-encrypt it.

Just to clarify: this happens for EVERY field even if is not useful/required to be encrypted. Correct?
Another thing to be written in the README :wink: 

> I am not sure about you question about secret configuration, it has to be in the odoo config file like stated in the readme.

Yes, I know. The reason why I mentioned it is because I find it more useful / easier to handle in such case (providing secret vars to an instance). I was just asking if one of the purpose of `server_environment_data_encryption` was to avoid using that secret var and how it should be used in relation to it.